File: subfinder.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"net/http"
	"os"
	"strings"
	"sync"
)

func getNSRecords(domain string) []string {
	nsRecords := []string{"ns1", "ns2", "ns3"}
	var nsIPs []string

	for _, ns := range nsRecords {
		nsIP, err := lookupNS(domain, ns)
		if err == nil {
			nsIPs = append(nsIPs, nsIP)
		}
	}

	return nsIPs
}

func lookupNS(domain string, ns string) (string, error) {
	nsAddress, err := lookupAddress(fmt.Sprintf("%s.%s", ns, domain))
	if err != nil {
		return "", err
	}
	return nsAddress, nil
}

func lookupAddress(host string) (string, error) {
	ips, err := net.LookupHost(host)
	if err != nil {
		return "", err
	}
	return ips[0], nil
}

func scanSubdomains(domain string, output chan string, wg *sync.WaitGroup) {
	defer wg.Done()

	subdomains, err := readLines("subdomains.txt")
	if err != nil {
		fmt.Println("Error reading subdomains:", err)
		return
	}

	for _, subdomain := range subdomains {
		url := fmt.Sprintf("http://%s.%s", subdomain, domain)
		resp, err := http.Get(url)
		if err == nil && resp.StatusCode == 200 {
			nsRecords := getNSRecords(domain)
			output <- fmt.Sprintf("Subdomain found: %s\nNS Records: %s\n\n", url, strings.Join(nsRecords, ", "))
		}
	}
}

func readLines(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func main() {
	domains, err := readLines("domains.txt")
	if err != nil {
		fmt.Println("Error reading domains:", err)
		return
	}

	var wg sync.WaitGroup
	output := make(chan string)

	for _, domain := range domains {
		wg.Add(1)
		go scanSubdomains(domain, output, &wg)
	}

	go func() {
		wg.Wait()
		close(output)
	}()

	resultFile, err := os.Create("output.txt")
	if err != nil {
		fmt.Println("Error creating output file:", err)
		return
	}
	defer resultFile.Close()

	for result := range output {
		resultFile.WriteString(result)
		fmt.Print(result)
	}
}

File: subfinder.py
import requests
from multiprocessing.dummy import Pool as ThreadPool

def scan_subdomain(domain):
    subdomains = []
    domain_results = []

    with open("subdomains.txt", "r") as subdomain_file:
        subdomains = subdomain_file.read().splitlines()

    for subdomain in subdomains:
        url = f"https://{subdomain}.{domain}"
        try:
            response = requests.get(url)

            if response.status_code in [200, 301, 400, 409, 502, 401]:
                server_info = response.headers.get('Server', 'N/A')
                print(f"Subdomain found: {url} | Status Code: {response.status_code} | Server: {server_info}\n")
                domain_results.append(url)

        except requests.RequestException:
            pass

    with open("output.txt", "w") as output_file:
        for result in domain_results:
            output_file.write(f"{result}\n")
            
    return domain_results

def main():
    domains = []

    with open("domains.txt", "r") as domain_file:
        domains = domain_file.read().splitlines()

    pool = ThreadPool(10)  # You can adjust the number of threads based on your system's capacity
    results = pool.map(scan_subdomain, domains)
    pool.close()
    pool.join()

if __name__ == "__main__":
    main()
